crates/zksync/core/src/vm/tracers/cheatcode.rs

```diff
@@ -235,7 +235,12 @@ impl<S: ReadStorage + StorageViewRecorder, H: HistoryMode> DynTracer<S, SimpleMe
         storage: StoragePtr<S>,
     ) {
         self.farcall_handler.track_call_actions(&state, &data);
-        let tx_tracking = self.farcall_handler.track_tx_execution(&state, &data, memory);
+
+        // We try to read calldata as late as possible since it's a very
+        // slow operation. When we do read it, we do it only once per call
+        // thru this cell
+        let calldata = std::cell::LazyCell::new(|| get_calldata(&state, memory));
+        let tx_tracking = self.farcall_handler.track_tx_execution(&state, &data, &calldata);
```

Comment:
How big of a difference did this make? I'd have preferred to do this in a separate PR for the following reasons:

We also have calldata retrieved also in `track_tx_execution` (it's even duplicated [here](https://github.com/matter-labs/foundry-zksync/blob/main/crates/zksync/core/src/vm/farcall.rs#L231) that can be removed). One can argue that we can have calldata returned as part of this, but then again tx execution tracking has a separate context.

Alternatively, we can define this functionality on farcall handler via more methods to reset/retrieve but that all depends on what is the impact exactly.

Reply:
that calldata retrieval you linked is removed in favor of this approach, passing down the cell.
I do agree that perhaps returning the calldata would be useful in an internal method instead of passing the cell around. Actually, I was even thinking, if we can detect / track which stack we are in we could keep track of the various calldata internally and return the right one so we would only read it once per frame (even tho I already feel that's the case)

Reply:
About 6% (for posterify)
We also resolved offline to just keep the cell, other solutions end up becoming more complicated due to the borrowing rules